app/src/ui/diff/image-diffs/image-diff.tsx

```diff
@@ -0,0 +1,116 @@
+import { NewImageDiff } from './new-image-diff'
+import { DeletedImageDiff } from './deleted-image-diff'
+
+import { ImageDiffType } from '../../../lib/app-state'
+import { Image, DiffHunk } from '../../../models/diff'
+import { TextDiff, ITextDiffUtilsProps } from '../text-diff'
+import { TabBar, TabBarType } from '../../tab-bar'
+
+/** The props for the Diff component. */
+interface IDiffProps extends ITextDiffUtilsProps {
+  /** The diff that should be rendered */
+  readonly previous?: Image
+  readonly current?: Image
+  readonly text?: string
+  readonly hunks?: ReadonlyArray<DiffHunk>
+
+  /** called when changing the type of image diff to display */
+  readonly onChangeImageDiffType: (type: ImageDiffType) => void
+
+  /** The type of image diff to display. */
+  readonly imageDiffType: ImageDiffType
+}
+
+/** A component which renders a diff for a file. */
+export class ImageDiff extends React.Component<IDiffProps, {}> {
+  private isModified = () => {
+    return this.props.current && this.props.previous
+  }
+  private onChangeDiffType = (index: number) => {
+    if (this.isModified()) {
+      this.props.onChangeImageDiffType(index)
+      return
+    }
+    this.props.onChangeImageDiffType(
+      index === 1 ? ImageDiffType.Text : ImageDiffType.TwoUp
+    )
+  }
+
+  private renderContent() {
+    if (
+      this.props.imageDiffType === ImageDiffType.Text &&
+      this.props.text &&
+      this.props.hunks
+    ) {
+      return (
+        <TextDiff
+          repository={this.props.repository}
+          readOnly={this.props.readOnly}
+          file={this.props.file}
+          text={this.props.text}
+          hunks={this.props.hunks}
+          onIncludeChanged={this.props.onIncludeChanged}
+        />
+      )
+    }
+    if (this.props.current && this.props.previous) {
+      return (
+        <ModifiedImageDiff
+          diffType={this.props.imageDiffType}
+          current={this.props.current}
+          previous={this.props.previous}
+        />
+      )
+    }
+
+    if (this.props.current) {
+      return <NewImageDiff current={this.props.current} />
+    }
+
+    if (this.props.previous) {
+      return <DeletedImageDiff previous={this.props.previous} />
+    }
+
+    return null
+  }
+
+  public render() {
+    const isModified = this.isModified()
+    const shouldRenderTabBar = this.props.text || isModified
+    let tabs
+    if (isModified) {
+      tabs = [
+        <span key="2-up">2-up</span>,
+        <span key="swipe">Swipe</span>,
+        <span key="onion">Onion Skin</span>,
+        <span key="diff">Difference</span>,
+      ]
+    } else {
+      tabs = [<span key="visual">Visual</span>]
+    }
+    if (this.props.text) {
+      tabs.push(<span key="text">Text</span>)
+    }
+    return (
+      <div className="panel image" id="diff">
+        {this.renderContent()}
+
+        {shouldRenderTabBar ? (
+          <TabBar
+            selectedIndex={
+              isModified
```

Comment:
We're doing some back-and-forth here mapping between `ImageDiffType` and the index in the tab bar list, which differs based on whether the diff is an image diff:

```ts
  private isModified = () => {
    return this.props.current && this.props.previous
  }
```

Before I go diving into this rabbit hole, I want to confirm what these rules are so I don't break how things currently work:

 - `this.isModified()` only looks for the images being set on `this.props` - is there a better name that we can use to differentiate between this "diffing both images" mode and the single diff mode?
 - `Visual` is a tab option but isn't a value of the `ImageDiffType` enum - is there a particular reason for that?

Reply:
* hum maybe, the `isModified` name comes from the git (as opposed to `added` or `removed`). Maybe we want to remove git terminology from this presentational component?

* because when the file is either added or removed, there is no type of visual diff. I didn't want to create another type in order to keep the previous `ImageDiffType` selected